In [26]:
!pip list | grep sci

scikit-image                       0.17.2
scikit-learn                       0.24.2
scipy                              1.5.2


In [10]:
# download the imbalance package
!pip install imblearn

     |████████████████████████████████| 206 kB 746 kB/s eta 0:00:01
     |████████████████████████████████| 24.9 MB 3.3 MB/s eta 0:00:01
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 0.23.2
    Uninstalling scikit-learn-0.23.2:
      Successfully uninstalled scikit-learn-0.23.2
Note: you may need to restart the kernel to use updated packages.


In [25]:
!pip upgrade scikit-learn

ERROR: unknown command "upgrade"


In [55]:
#importing all the important libraries
import pandas as pd

In [56]:
#loading the csv file
data = pd.read_csv('car_evaluation.csv')
data.head()

buying  maint  doors  persons lug_boot safety outcome
0  vhigh  vhigh      2        2    small    low   unacc
1  vhigh  vhigh      2        2    small    med   unacc
2  vhigh  vhigh      2        2    small   high   unacc
3  vhigh  vhigh      2        2      med    low   unacc
4  vhigh  vhigh      2        2      med    med   unacc

In [57]:
# shape of the data
data.shape

(1728, 7)

In [58]:
# lets the target value data
data.outcome.value_counts()

unacc    1210
acc       384
good       69
vgood      65
Name: outcome, dtype: int64

In [59]:
# seprating the indepedndent and the dependent variable
X = data.iloc[:,:-1]
y = data.outcome
X.head()

buying  maint  doors  persons lug_boot safety
0  vhigh  vhigh      2        2    small    low
1  vhigh  vhigh      2        2    small    med
2  vhigh  vhigh      2        2    small   high
3  vhigh  vhigh      2        2      med    low
4  vhigh  vhigh      2        2      med    med

In [60]:
from sklearn.preprocessing import LabelEncoder
enc = LabelEncoder()
X.loc[:,['buying','maint','lug_boot','safety']] = \
X.loc[:,['buying','maint','lug_boot','safety']].apply(enc.fit_transform)
X.head()

buying  maint  doors  persons  lug_boot  safety
0       3      3      2        2         2       1
1       3      3      2        2         2       2
2       3      3      2        2         2       0
3       3      3      2        2         1       1
4       3      3      2        2         1       2

In [61]:
from sklearn.model_selection import train_test_split
X_train,X_test, y_train,y_test = \
train_test_split(X,y,test_size=0.3,random_state=10)

In [62]:
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier()
model.fit(X_train,y_train)
y_predict = model.predict(X_test)

In [63]:
from sklearn.metrics import accuracy_score , classification_report
print(accuracy_score(y_test,y_predict))
print(pd.crosstab(y_test,y_predict))
print(classification_report(y_test,y_predict))

0.928709055876686
col_0    acc  good  unacc  vgood
outcome                         
acc       89     2     11      0
good      10    10      1      0
unacc      5     0    366      0
vgood      5     2      1     17
              precision    recall  f1-score   support

         acc       0.82      0.87      0.84       102
        good       0.71      0.48      0.57        21
       unacc       0.97      0.99      0.98       371
       vgood       1.00      0.68      0.81        25

    accuracy                           0.93       519
   macro avg       0.87      0.75      0.80       519
weighted avg       0.93      0.93      0.93       519



lets try to calculate the accuracy of each labels

acc - 89/102 - 0.87

good - 10/21 - 0.47

unacc - 366/371 - 0.98

vgood - 17/25 - 0.68

In [107]:
from imblearn.over_sampling import SMOTE
smote = SMOTE(random_state=104)

In [108]:
X_train_smote, y_train_smote = smote.fit_resample(X_train.astype('float'),y_train)

In [109]:
X_train_smote

buying     maint     doors  persons  lug_boot  safety
0     0.000000  0.000000  3.000000      5.0  0.000000     0.0
1     1.000000  3.000000  2.000000      5.0  1.000000     1.0
2     2.000000  3.000000  4.000000      5.0  2.000000     1.0
3     0.000000  1.000000  3.000000      5.0  2.000000     0.0
4     0.000000  1.000000  4.000000      4.0  2.000000     0.0
...        ...       ...       ...      ...       ...     ...
3351  1.828910  1.000000  4.000000      5.0  1.000000     0.0
3352  1.147635  1.852365  5.000000      5.0  0.000000     0.0
3353  1.128912  2.000000  4.000000      4.0  1.000000     0.0
3354  1.552501  0.552501  2.552501      4.0  0.000000     0.0
3355  1.000000  0.969043  5.000000      4.0  0.030957     0.0

[3356 rows x 6 columns]

In [110]:
from collections import Counter
print("Before SMOTE :" , Counter(y_train))
print("After SMOTE :" , Counter(y_train_smote))

Before SMOTE : Counter({'unacc': 839, 'acc': 282, 'good': 48, 'vgood': 40})
After SMOTE : Counter({'acc': 839, 'unacc': 839, 'vgood': 839, 'good': 839})


In [111]:
model.fit(X_train_smote,y_train_smote)
y_predict = model.predict(X_test)
print(accuracy_score(y_test,y_predict))
print(pd.crosstab(y_test,y_predict))
print(classification_report(y_test,y_predict))

0.928709055876686
col_0    acc  good  unacc  vgood
outcome                         
acc       90     7      3      2
good       1    20      0      0
unacc     20     1    349      1
vgood      1     1      0     23
              precision    recall  f1-score   support

         acc       0.80      0.88      0.84       102
        good       0.69      0.95      0.80        21
       unacc       0.99      0.94      0.97       371
       vgood       0.88      0.92      0.90        25

    accuracy                           0.93       519
   macro avg       0.84      0.92      0.88       519
weighted avg       0.94      0.93      0.93       519



lets try to calculate the accuracy of each labels

acc - 94/104 - 0.90

good - 20/21 - 0.95

unacc - 355/371 - 0.95

vgood - 23/25 - 0.92